File: SortVision/public/code/heap/rust/heapSort.rs
/// Heap Sort Implementation in Rust
///
/// # Overview
/// Sorts a mutable slice in-place using the Heap Sort algorithm.
/// It builds a max-heap and repeatedly extracts the maximum element.
///
/// # Time Complexity
/// - Worst/Average/Best: O(n log n)
///
/// # Space Complexity
/// - O(1): In-place sorting
///
/// # Arguments
/// * `arr` - A mutable slice of elements that implement the `Ord` trait
///
/// # Constraints
/// * Elements must implement the `Ord` trait
///
/// # Example
/// ```
/// let mut arr = vec![4, 10, 3, 5, 1];
/// heap_sort(&mut arr);
/// assert_eq!(arr, vec![1, 3, 4, 5, 10]);
/// ```
pub fn heap_sort<T: Ord>(arr: &mut [T]) {
    let n = arr.len();

    // Step 1: Build a max heap
    for i in (0..n / 2).rev() {
        heapify(arr, n, i);
    }

    // Step 2: Extract elements from heap
    for i in (1..n).rev() {
        arr.swap(0, i);           // Move current root to end
        heapify(arr, i, 0);       // Heapify reduced heap
    }
}

/// Converts a subtree rooted at index `i` into a max heap.
///
/// # Arguments
/// * `arr` - The array to be heapified
/// * `n` - Size of the heap
/// * `i` - Index of the root of the subtree
///
/// # Notes
/// - This function assumes the binary trees rooted at the left and right child of `i` are already heaps.

pub fn heapify<T: Ord>(arr: &mut [T], n: usize, i: usize) {
    let mut largest = i;
    let left = 2 * i + 1;
    let right = 2 * i + 2;

    if left < n && arr[left] > arr[largest] {
        largest = left;
    }

    if right < n && arr[right] > arr[largest] {
        largest = right;
    }

    if largest != i {
        arr.swap(i, largest);
        heapify(arr, n, largest);
    }
}

/// Demonstrates example usage of heap_sort
fn main() {
    let mut data = vec![4, 10, 3, 5, 1];
    println!("Before sorting: {:?}", data);
    heap_sort(&mut data);
    println!("After sorting:  {:?}", data);

    let mut test_cases = vec![
        vec![],
        vec![42],
        vec![9, 8, 7],
        vec![2, 1, 3, 5, 4],
        vec![5, 5, 5, 2, 2],
    ];

    for (i, case) in test_cases.iter_mut().enumerate() {
        heap_sort(case);
        println!("Test {} → {:?}", i + 1, case);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty() {
        let mut arr: Vec<i32> = vec![];
        heap_sort(&mut arr);
        assert_eq!(arr, vec![]);
    }

    #[test]
    fn test_single() {
        let mut arr = vec![1];
        heap_sort(&mut arr);
        assert_eq!(arr, vec![1]);
    }

    #[test]
    fn test_sorted() {
        let mut arr = vec![1, 2, 3, 4, 5];
        heap_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_reverse() {
        let mut arr = vec![5, 4, 3, 2, 1];
        heap_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_duplicates() {
        let mut arr = vec![4, 1, 3, 1, 4];
        heap_sort(&mut arr);
        assert_eq!(arr, vec![1, 1, 3, 4, 4]);
    }
}

File: SortVision/public/code/heap/swift/heapSort.swift
/*
Heap Sort Algorithm in Swift

Time Complexity:
- Worst Case: O(n log n)
- Average Case: O(n log n)
- Best Case: O(n log n)

Space Complexity: O(1) (in-place)

Features:
- Builds a max-heap from the array
- Repeatedly swaps max to end, shrinks heap
- Handles edge cases (empty array, single element, duplicates)
*/

import Foundation

class HeapSort {
    // Public method to sort an array using heap sort
    static func sort(_ arr: inout [Int]) {
        let n = arr.count
        buildMaxHeap(&arr, n)

        for i in stride(from: n - 1, through: 1, by: -1) {
            arr.swapAt(0, i)
            heapify(&arr, i, 0)
        }
    }

    // Builds a max heap from the array
    private static func buildMaxHeap(_ arr: inout [Int], _ n: Int) {
        for i in stride(from: n / 2 - 1, through: 0, by: -1) {
            heapify(&arr, n, i)
        }
    }

    // Ensures the subtree rooted at index `i` is a max heap
    private static func heapify(_ arr: inout [Int], _ n: Int, _ i: Int) {
        var largest = i
        let left = 2 * i + 1
        let right = 2 * i + 2

        if left < n && arr[left] > arr[largest] {
            largest = left
        }

        if right < n && arr[right] > arr[largest] {
            largest = right
        }

        if largest != i {
            arr.swapAt(i, largest)
            heapify(&arr, n, largest)
        }
    }
}


func testHeapSort() {
    var test1: [Int] = []
    var test2: [Int] = [42]
    var test3: [Int] = [5, 3, 8, 4, 1]
    var test4: [Int] = [7, 2, 2, 9, 5]

    HeapSort.sort(&test1)
    HeapSort.sort(&test2)
    HeapSort.sort(&test3)
    HeapSort.sort(&test4)

    assert(test1 == [])
    assert(test2 == [42])
    assert(test3 == [1, 3, 4, 5, 8])
    assert(test4 == [2, 2, 5, 7, 9])

    print("All tests passed!")
}

testHeapSort()
